### PR TITLE
Header components

### LINKED ISSUE #331: [BUG] Components are either not ready or broken
**Describe the bug**
I've tried to ensemble this homepage template with the design-kit: https://italia.github.io/design-comuni-prototipi/esempi/bootstrap-italia/template-homepage.html

The following list is what I've tried so far:
* [HeaderNav](https://italia.github.io/design-react-kit/?path=/story/componenti-headernav--esempio) is of no use as it is
* [Navbar](https://italia.github.io/design-react-kit/?path=/story/componenti-navbar--esempi) works but the documentation is of no use. I had to refer to the original reactstrap documentation to build something
* [Navbar Dropdown](https://italia.github.io/design-react-kit/?path=/story/componenti-navbar--con-dropdown) the documentation is of no use.
* [Button](https://italia.github.io/design-react-kit/?path=/story/componenti-button--esempi) works but the documentation has to be improved (props table is of no use)
* SVG icons - there's no documented way to use them.
* [Card](https://italia.github.io/design-react-kit/?path=/story/componenti-cards--simple-card) semantic is completely messed up. This should follow the reactstrap semantic for consistency and not reinvent it with a waterfall of props. The problem of the current design is that is does not provide any composable way to combine inner parts of the card compared to the original semantic.

### PR DESCRIPTION
**Pull request by dej611**
Fixes #331

#### PR Checklist

- [x] My branch is up-to-date with the Upstream `next` branch.
- [x] The unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

#### Short description of what this resolves:

This PR introduces the component Headers with its different variants as described in the official [BI documentation](https://italia.github.io/bootstrap-italia/docs/menu-di-navigazione/header/).

#### Changes proposed in this pull request:

**Components**

- [x] Headers wrapper component
- [x] Header component with different types and themes
- [x] Header Brand component
- [x] Header Content
- [x] Header Context
- [x] Header Link Zone - for Slim header
- [x] Header Right Zone - for Slim header
- [x] Header Search - for Center header
- [x] Header Toggler - custom version of Navbar toggler for headers
- [x] Nav - custom version of Nav with headers switch
  - [x] NavHeader - custom subtype
- [x] Megamenu - TBD

**Storybook**
- [x] Introduction
- Slim Header
  - [x] basic
  - [x] basic doc
  - [x] basic prop tables
  - [x] light
  - [x] light doc
  - [x] light prop tables
- Center Header
  - [x] basic
  - [x] basic doc
  - [x] basic prop tables
  - [x] light
  - [x] light doc
  - [x] light prop tables
  - [x] small
  - [x] small doc
  - [x] small prop tables
- Nav Header
  - [x] basic
  - [x] basic doc
  - [x] basic prop tables
  - [x] light-desk
  - [x] light-desk doc
  - [x] light-desk prop tables
  - [x] dark-mobile
  - [x] dark-mobile doc
  - [x] dark-mobile prop tables
- Complete Header
  - [x] basic
  - [x] basic doc
  - [x] basic prop tables
  - [x] light
  - [x] light doc
  - [x] light prop tables

**Various**
- [x] Remove eslint warnings

### CODE REVIEW

**Comment by dej611** [comment]
Components documentation has been added, but docgen is not picking that up when the component is a class.

**Comment by dej611** [comment]
Dropdowns need to be sorted:
- `Slim` Header needs some custom styling and structure.
- `Nav` Header needs special styling.

Perhaps worth investigating in a separate PR.

File: src/components/Header/HeaderRightZone.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    static propTypes = propTypes;
+    static defaultProps = defaultProps;
+    static contextType = HeaderContext;
+
+    render() {
+        const { className, ...attributes } = this.props;
+        const { type } = this.context;
+        const classes = classNames(className, {
+            'header-slim-right-zone': type === SLIM,
```
Comment (francescozaia) [inline src/components/Header/HeaderRightZone.js]: Da Bootstrap Italia 1.3.6 è possibile usare `it-header-slim-right-zone`.

Comment (dej611) [reply src/components/Header/HeaderRightZone.js]: Aggiunto anche la classe per BI >= 1.3.6

**Comment by dej611** [comment]
Si i dropdown sono un capitolo a parte. Posso inserirli senza "componentizzarli" esattamente come in Bootstrap Italia, oppure fare una PR a parte per mettere in linea il componente `Dropdown` con lo stile BI

File: stories/Header/CompleteHeader/basic.js [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    render() {
+        return (
+            // eslint-disable-next-line react/prop-types
+            <Header type="slim" theme={this.props.theme}>
+                <HeaderContent>
+                    <HeaderBrand>Ente appartenenza/Owner</HeaderBrand>
+                    <HeaderLinkZone>
+                        <HeaderToggler onClick={() => this.setState({ collapsed: !this.state.collapsed })}>
+                            <span>Ente appartenenza/Owner</span>
+                            <Icon icon="it-expand" />
```
Comment (sanyamdogra) [inline stories/Header/CompleteHeader/basic.js]: The cursor behaves like it's an input field. May be due to `href` or type in `HeaderLinkZone. I don't think it should change the cursor to `|`.

**Comment by francescozaia** [comment]
> Si i dropdown sono un capitolo a parte. Posso inserirli senza "componentizzarli" esattamente come in Bootstrap Italia, oppure fare una PR a parte per mettere in linea il componente `Dropdown` con lo stile BI

Riguardo questo appunto, prima di mergiare preferiremmo fosse completamente funzionante, o in alternativa spiegare come farla funzionare correttamente anche in caso di dispositivi che mostrino l'hamburger menu.

**Comment by dej611** [comment]
headernav-mobile

**Comment by dej611** [comment]
Things to fix:
* [ ] In the Header nav side menu, the dropdown submenu is broken
* [ ] In the Slim header nav responsive for the Homepage, the language + profile section should be on the right handside

**Comment by dej611** [comment]
Did a dig on the issue within the dropdown submenu issue here:
dropdown-bug

The root of the problem comes from `react-popper` and the `popper` library used internally in `reactstrap`.
The `popper` library sets the `position` value of the dropdown to `absolute` to position the list of items passed, with the effect of not pushing the item below.
It would be possible to bypass that in theory, but `react-popper` [overwrites the `modifiers` `applyStyles`](https://github.com/FezVrasta/react-popper/blob/9763593b63bb8f380855c9b5683cda2886e90a60/src/Popper.js#L110) passed thru `reactstrap` in a way that makes unreachable any `positioning` styling attribute.

There's a pending PR for that which MAY solve the problem.

As a quick fix I can now remove the dropdown from the header nav, but it is probably worth investigating a different way to have the dropdown instead of the `Dropdown` component provided by `reactstrap` in this scenario.

**Comment by francescozaia** [comment]
Reminder: this issue with the slim header is going to be fixed with an update on Bootstrap Italia: https://github.com/italia/bootstrap-italia/pull/398

Screenshot 2019-10-01 at 12 41 16

**Pull request merged**